clippy_lints/src/matches/match_on_vec_items.rs

```
@@ -1,27 +1,66 @@
-use clippy_utils::diagnostics::span_lint_and_sugg;
-use clippy_utils::source::snippet;
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::is_wild;
+use clippy_utils::msrvs::{self, Msrv};
+use clippy_utils::source::{indent_of, snippet, snippet_with_applicability};
 use clippy_utils::ty::{is_type_diagnostic_item, is_type_lang_item};
 use rustc_errors::Applicability;
-use rustc_hir::{Expr, ExprKind, LangItem};
+use rustc_hir::{Arm, Expr, ExprKind, LangItem};
 use rustc_lint::LateContext;
-use rustc_span::sym;
+use rustc_span::{Span, sym};
 
 use super::MATCH_ON_VEC_ITEMS;
 
-pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, scrutinee: &'tcx Expr<'_>) {
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, scrutinee: &'tcx Expr<'_>, arms: &'tcx [Arm<'_>], msrv: &Msrv) {
     if let Some(idx_expr) = is_vec_indexing(cx, scrutinee)
         && let ExprKind::Index(vec, idx, _) = idx_expr.kind
     {
-        // FIXME: could be improved to suggest surrounding every pattern with Some(_),
-        // but only when `or_patterns` are stabilized.
-        span_lint_and_sugg(
+        span_lint_and_then(
             cx,
             MATCH_ON_VEC_ITEMS,
             scrutinee.span,
             "indexing into a vector may panic",
-            "try",
-            format!("{}.get({})", snippet(cx, vec.span, ".."), snippet(cx, idx.span, "..")),
-            Applicability::MaybeIncorrect,
+            |diag| {
+                if msrv.meets(msrvs::OR_PATTERNS) {
+                    let mut app = Applicability::MachineApplicable;
```

I don't think this is right to use `MachineApplicable` here. You are silently converting a `panic!()` into calling the default alternative, that is certainly not equivalent.

For example, calling the following function with `&vec![1, 2, 3]` will panic before the "fix", and will print "not 2" after, as if an element had been correctly retrieved and contained a value different from 2.

```rust
fn f(v: &Vec<u32>) {
    match v[4] {
        2 => println!("2"),
        _ => println!("not 2"),
    };
}
```

---

It was certainly premature to change applicability to `MachineApplicable` here. Then, should it be `MaybeIncorrect` or `Unspecified`? In any case, the original implementation set applicability to `MaybeIncorrect` without modifying each arm, so I think this needs to be changed.

---

My position here is that we should not even make a suggestion as it definitely changes the behavior. I'll open a thread on Zulip to discuss that.